### PR TITLE
[MRG] Added Jensen Shannon metric and dendrogram visualization

### PR DESCRIPTION
**Pull request by parulsethi**
Adds Jensen-Shannon distance metric and topic dendrogram-heatmap visualization notebook.

### CODE REVIEW

**Comment by piskvorky** [comment]
Hanging indent everywhere please -- vertical indent makes the code hard to read and hard to maintain.

**Comment by piskvorky** [comment]
@parulsethi It's Jensen-Shannon (after Johan Jensen and Claude Shannon).

**Comment by parulsethi** [comment]
@piskvorky Can you please point to which line should be changed to hanging indent? As if i understand correctly hanging/vertical indent is used in case of long arguments? and there weren't any in this PR so I didn't use any line breaks in code

**Comment by piskvorky** [comment]
Unfortunately Github says the notebook is too large and its diff cannot be displayed, so I cannot review directly. But the indentation is broken around `annotation_html =`, `update(` and a few other places.

There are also a few minor code style inconsistencies, like spaces around `,` or `:`. @menshikh-iv will direct you :)

@menshikh-iv is there a way to somehow review the notebook? I'd like to point out some constructs which, while not wrong, could be improved a little for better readability (like `map(int(`), to be more "Pythonic".

**Comment by parulsethi** [comment]
@piskvorky I'll remove the cell outputs once the notebook is complete. It will then have a much smaller diff containing only the input code cells which could be displayed in Github for code review.

**Comment by piskvorky** [comment]
@parulsethi let's fix that `Jenson`, before it propagates too far.

**Comment by parulsethi** [comment]
I've removed the plotly code and cell outputs for smaller diff on github, will add it back after the code review of notebook is complete. If needed, the output visualization can be seen [here](https://github.com/parulsethi/gensim/blob/b9e7ab54dde98438b0e4f766ee764b81af704367/docs/notebooks/Topic_dendrogram.ipynb) in a previous commit.

@piskvorky corrected the name in a notebook comment it was being used in. Elsewhere it's the function name.

**Comment by piskvorky** [comment]
The name is incorrect and needs to be fixed.

**Comment by parulsethi** [comment]
Oh, the spelling. Sorry, misunderstood the first time

**Comment by menshikh-iv** [comment]
@piskvorky About notebook - [nbdime](https://github.com/jupyter/nbdime) good tool for view diffs between notebooks, but this tool doesn't integrate with github.

Also, we can use `nbconvert`, but this again does not suit us. Probably it makes sense to talk with tech GitHub support and discuss their plans to improve the work with notebooks.

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from gensim.models.ldamodel import LdaModel\n",
+    "from gensim.corpora import Dictionary, MmCorpus\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: Unused imports: `gensim.corpora.MmCorpus`,  `scipy.spatial.distance.squareform`, `plotly.figure_factory`

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    "def preprocess(row):\n",
+    "    return strip_punctuation(remove_stopwords(row.lower()))\n",
+    "    \n",
+    "df_fake['text'] = df_fake['text'].apply(preprocess)\n",
+    "\n",
+    "# Convert data to required input format by LDA\n",
+    "texts = []\n",
+    "for line in df_fake.text:\n",
+    "    lowered = line.lower()\n",
+    "    words = re.findall(r'\\w+', lowered, flags = re.UNICODE | re.LOCALE)\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: Unnecessary  spaces `flags = re.UNICODE | re.LOCALE`

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true,
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "from gensim.matutils import jensen_shannon\n",
+    "\n",
+    "from random import sample\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: Unused imports: `random.sample`, `scipy`

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# get text annotations\n",
+    "annotation = text_annotation(topic_dist, topic_terms, n_ann_terms)\n",
+    "\n",
+    "# Plot dendrogram\n",
+    "dendro = create_dendrogram(topic_dist, distfun=js_dist, labels=range(1, 36), annotation=annotation)\n",
+    "dendro['layout'].update({'width':1000, 'height':600})\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: Spaces after `:`

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# get text annotations\n",
+    "annotation = text_annotation(topic_dist, topic_terms, n_ann_terms)\n",
+    "\n",
+    "# Initialize figure by creating upper dendrogram\n",
+    "figure = create_dendrogram(topic_dist, distfun=js_dist, labels = range(1, 36), annotation=annotation)\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: no needed spaces `labels = range(1, 36)`

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# get distance matrix and it's topic annotations\n",
+    "mdiff, annotation = lda_fake.diff(lda_fake, distance=\"jensen_shannon\", normed=False)\n",
+    "\n",
+    "# get reordered topic list\n",
+    "dendro_leaves = figure['layout']['xaxis']['ticktext']\n",
+    "dendro_leaves = list(map(int, dendro_leaves-1))\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: What's type of dendro_leaves? Maybe you want to `[int(dendro_leaves  - 1)]`

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    "\n",
+    "# reorder distance matrix\n",
+    "heat_data = mdiff[dendro_leaves,:]\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: space after `,` (and below)

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# heatmap annotation\n",
+    "annotation_html = [[\"+++ {}<br>--- {}\".format(\", \".join(int_tokens), \", \".join(diff_tokens))\n",
+    "                    for (int_tokens, diff_tokens) in row]\n",
+    "                    for row in annotation]\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: move to previous line

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+    "\n",
+    "# plot heatmap of distance matrix\n",
+    "heatmap = go.Data([\n",
+    "    go.Heatmap(\n",
+    "        z = heat_data,\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: No needed spaces (and below)

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    "    )\n",
+    "])\n",
+    "\n",
+    "heatmap[0]['x'] = figure['layout']['xaxis']['tickvals']\n",
+    "heatmap[0]['y'] = figure['layout']['xaxis']['tickvals']\n",
+    "\n",
+    "# Add Heatmap Data to Figure\n",
+    "figure['data'].extend(heatmap)\n",
+    "\n",
+    "dendro_leaves = [x+1 for x in dendro_leaves]\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: Missing whitespace `x+1`

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+    "\n",
+    "# Edit Layout\n",
+    "figure['layout'].update({'width':800, 'height':800,\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: missing spaces after `:` (and below)

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+    "                         })\n",
+    "\n",
+    "# Edit xaxis\n",
+    "figure['layout']['xaxis'].update({'domain': [.25, 1],\n",
+    "                                  'mirror': False,\n",
+    "                                  'showgrid': False,\n",
+    "                                  'showline': False,\n",
+    "                                  \"showticklabels\": True, \n",
+    "                                  \"tickmode\": \"array\",\n",
+    "                                  \"ticktext\" : dendro_leaves,\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: Unnecessary space (before `:`) and below

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+    "                                  'zeroline': False,\n",
+    "                                  'ticks':\"\"})\n",
+    "# Edit yaxis\n",
+    "figure['layout']['yaxis'].update({'domain': [0, 0.75],\n",
+    "                                  'mirror': False,\n",
+    "                                  'showgrid': False,\n",
+    "                                  'showline': False,\n",
+    "                                  \"showticklabels\": True, \n",
+    "                                  \"tickmode\": \"array\",\n",
+    "                                  \"ticktext\" : dendro_leaves,\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: Unnecessary space (before `:`) and below

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+    "                                  'zeroline': False,\n",
+    "                                  'ticks': \"\"})\n",
+    "# Edit yaxis2\n",
+    "figure['layout'].update({'yaxis2':{'domain':[0.75, 1],\n",
+    "                                   'mirror': False,\n",
+    "                                   'showgrid': False,\n",
+    "                                   'showline': False,\n",
+    "                                   'zeroline': False,\n",
+    "                                   'showticklabels': False,\n",
+    "                                   'ticks':\"\"}})\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Topic_dendrogram.ipynb]: missing space

**Comment by menshikh-iv** [comment]
@parulsethi  please fix PEP things, return all images and resolve merge conflict. LGTM for me

File: docs/notebooks/Topic_dendrogram.ipynb [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+    "topic_dist = lda_fake.state.get_lambda()\n",
+    "\n",
+    "# get topic terms\n",
+    "num_words = 300\n",
+    "topic_terms = [{w for (w, _) in lda_fake.show_topic(topic, topn=num_words)} for topic in range(topic_dist.shape[0])]\n",
+    "\n",
+    "# no. of terms to display in annotation\n",
+    "n_ann_terms = 10\n",
+    "\n",
+    "# use Jenson-Shannon distance metric in dendrogram\n",
```
Comment (piskvorky) [inline docs/notebooks/Topic_dendrogram.ipynb]: One more "Jenson".

File: docs/notebooks/Topic_dendrogram.ipynb (hunk 6 above)
Comment (parulsethi) [inline docs/notebooks/Topic_dendrogram.ipynb]: It's a list of int values where every value needs to be subtracted by 1. Simplified this using list comprehension now

**Comment by menshikh-iv** [comment]
Good job @parulsethi :1st_place_medal:

**Pull request merged**